### PR TITLE
Avoid requesting for advanced (far) blocks.

### PR DESCRIPTION
**Pull request by casiojapi**
## Description

In the processNewBlockHashesMessage method we could probably avoid requesting too advanced blocks (for example, if node is on early stages of a full-sync).

The main idea of this change is to avoid requesting these "far" blocks which will be later discarded. And by doing that, we can optimize the "processNewBlockHashesMessage" method.

This way we could save some unnecessary load in both nodes.
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Requires Activation Code (Hard Fork)

* **Other information**:

### CODE REVIEW

**Comment by casiojapi** [comment]
pipeline:run

**Comment by casiojapi** [comment]
pipeline:run

**Comment by casiojapi** [comment]
pipeline:run

File: rskj-core/src/test/java/co/rsk/net/NodeBlockProcessorTest.java [hunk 1]
```diff
@@ -1082,4 +1084,66 @@ void failIfProcessBlockHeadersRequestCountHigher()  {
...
+    void processNewBlockHashesMessageOnlyRequestsNonAdvancedBlock() {
+        final NetBlockStore store = new NetBlockStore();
+        final Peer sender = mock(SimplePeer.class);
+        final BlockNodeInformation nodeInformation = mock(BlockNodeInformation.class);
+        final SyncConfiguration syncConfiguration = SyncConfiguration.IMMEDIATE_FOR_TESTING;
+        final Blockchain blockchain = new BlockChainBuilder().ofSize(0);
+
+        TestSystemProperties config = new TestSystemProperties();
+        BlockSyncService blockSyncService = new BlockSyncService(config, store, blockchain, nodeInformation, syncConfiguration, DummyBlockValidator.VALID_RESULT_INSTANCE);
+        final NodeBlockProcessor processor = new NodeBlockProcessor(store, blockchain, nodeInformation, blockSyncService, syncConfiguration);
```
Comment (asoto-iov) [inline rskj-core/src/test/java/co/rsk/net/NodeBlockProcessorTest.java:1098]: (minor thing) I'd extract this logic to a method as it seems the same code is used in other tests just to create the processor.

File: rskj-core/src/test/java/co/rsk/net/NodeBlockProcessorTest.java [hunk 2]
```diff
@@ -1082,4 +1084,66 @@ void failIfProcessBlockHeadersRequestCountHigher()  {
...
+
+        final long advancedBlockNumber = (long) syncConfiguration.getChunkSize() * syncConfiguration.getMaxSkeletonChunks() + blockchain.getBestBlock().getNumber() + 1;
+
+        BlockIdentifier advancedBlockIdentifier = new BlockIdentifier(new byte[32], advancedBlockNumber);
+
+        // create list of block identifiers with advanced and non-advanced blocks
```
Comment (asoto-iov) [inline rskj-core/src/test/java/co/rsk/net/NodeBlockProcessorTest.java:1136]: This comment does not belong to here :D

File: rskj-core/src/test/java/co/rsk/net/NodeBlockProcessorTest.java [hunk 3]
```diff
@@ -1082,4 +1084,66 @@ void failIfProcessBlockHeadersRequestCountHigher()  {
...
+        blockIdentifierList.add(nonAdvancedBlockIdentifier);
+        blockIdentifierList.add(advancedBlockIdentifier);
+
+        NewBlockHashesMessage message = new NewBlockHashesMessage(blockIdentifierList);
+
+        processor.processNewBlockHashesMessage(sender, message);
+
+        // verify that only the non-advanced block is requested and added to the node information
+        verify(sender, times(1)).sendMessage(any(GetBlockMessage.class));
+        verify(nodeInformation, times(1)).addBlockToNode(any(), any());
```
Comment (asoto-iov) [inline rskj-core/src/test/java/co/rsk/net/NodeBlockProcessorTest.java:1117]: ArgumentCaptor could be used here and check if the block added is the right one, as into this test is not truly checked. Having that maybe the next unit test wouldn't be needed.

Comment (casiojapi) [reply rskj-core/src/test/java/co/rsk/net/NodeBlockProcessorTest.java:1117]: That's a pretty useful idea. Just Implemented it and now with one test is enough. Thanks!!!

**Comment by casiojapi** [comment]
pipeline:run

**Comment by casiojapi** [comment]
pipeline:run

**Comment by casiojapi** [comment]
pipeline:run

**Comment by casiojapi** [comment]
pipeline:run

**Review by Vovchyk — DISMISSED** [review]

**Comment by casiojapi** [comment]
pipeline:run

**Comment by casiojapi** [comment]
pipeline:run

**Comment by casiojapi** [comment]
pipeline:run

**Review by Vovchyk — APPROVED** [review]

**Pull request merged**